sqlx-core/src/postgres/types/decimal.rs

```diff
@@ -0,0 +1,414 @@
+            return Ok(0u64.into());
+        }
+
+        let sign = match sign {
+            PgNumericSign::Positive => Sign::Plus,
+            PgNumericSign::Negative => Sign::Minus,
+        };
+
+        // weight is 0 if the decimal point falls after the first base-10000 digit
+        let scale = (digits.len() as i64 - weight as i64 - 1) * 4;
+
+        // no optimized algorithm for base-10 so use base-100 for faster processing
+        let mut cents = Vec::with_capacity(digits.len() * 2);
+        for digit in &digits {
+            cents.push((digit / 100) as u8);
+            cents.push((digit % 100) as u8);
+        }
+
+        let bigint = BigInt::from_radix_be(sign, &cents, 100)
+            .ok_or("PgNumeric contained an out-of-range digit")?;
+
+        match bigint.to_i128() {
+            Some(num) => Ok(Decimal::from_i128_with_scale(num, scale as u32)),
+            None => Err("Decimal's integer part out of range.".into()),
+        }
+    }
+}
+
+impl TryFrom<&'_ Decimal> for PgNumeric {
+    type Error = BoxDynError;
+
+    fn try_from(decimal: &Decimal) -> Result<Self, BoxDynError> {
+        if decimal.is_zero() {
+            return Ok(PgNumeric::Number {
+                sign: PgNumericSign::Positive,
+                scale: 0,
+                weight: 0,
+                digits: vec![],
+            });
+        }
+
+        let scale = decimal.scale() as u16;
+
+        let groups_diff = scale & 0x3;
+
+        let s = decimal.serialize();
+        let mut mantissa = [
+            u32::from_le_bytes([s[4], s[5], s[6], s[7]]),
+            u32::from_le_bytes([s[8], s[9], s[10], s[11]]),
+            u32::from_le_bytes([s[12], s[13], s[14], s[15]]),
+            0u32,
+        ];
+
+        if groups_diff > 0 {
+            let remainder = 4 - groups_diff;
+            let power = 10u32.pow(u32::from(remainder));
+            let mut overflow = 0;
+
+            for b in mantissa.iter_mut() {
+                let (lo, hi) = {
+                    let result = u64::from(*b) * u64::from(power) + u64::from(overflow);
+                    let hi = ((result >> 32) & 0xFFFF_FFFF) as u32;
+                    let lo = (result & 0xFFFF_FFFF) as u32;
+                    (lo, hi)
+                };
+
+                *b = lo;
+                overflow = hi;
+            }
+        }
+
+        let mut digits = Vec::with_capacity(8);
+
+        while !mantissa.iter().all(|b| *b == 0) {
+            let mut remainder = 0u32;
+            let divisor = 10000u64;
+
+            for part in mantissa.iter_mut().rev() {
+                let temp = (u64::from(remainder) << 32) + u64::from(*part);
+                remainder = (temp % divisor) as u32;
+                *part = (temp / divisor) as u32;
+            }
+
+            digits.push(remainder as i16)
+        }
+
+        digits.reverse();
+
+        let digits_after_decimal = (scale + 3) as u16 / 4;
+        let weight = digits.len() as i16 - digits_after_decimal as i16 - 1;
```

Comment:
There are some interesting math things going on here -- comments would be helpful for future maintainability.